pkg/policy/mapstate.go

```diff
@@ -314,25 +406,190 @@ func (keys MapState) denyPreferredInsertWithChanges(newKey Key, newEntry MapStat
+	// Merging owners from the new entry to the existing one has no datapath impact so we skip
 	// adding anything to 'adds' here.
 	if oldEntry, exists := keys[key]; exists && (oldEntry.IsRedirectEntry() || oldEntry.IsDeny) {
-		oldEntry.MergeSelectors(&entry)
+		oldEntry.MergeReferences(&entry)
 		keys[key] = oldEntry
 		return
 	}
 	// Otherwise write the entry to the map
 	keys.addKeyWithChanges(key, entry, adds, deletes)
 }
 
+var visibilityDerivedFromLabels = labels.LabelArray{
+	labels.NewLabel(LabelKeyPolicyDerivedFrom, LabelVisibilityAnnotation, labels.LabelSourceReserved),
+}
+
+var visibilityDerivedFrom = labels.LabelArrayList{visibilityDerivedFromLabels}
+
+// AddVisibilityKeys adjusts and expands PolicyMapState keys
+// and values to redirect for visibility on the port of the visibility
+// annotation while still denying traffic on this port for identities
+// for which the traffic is denied.
+//
+// Datapath lookup order is, from highest to lowest precedence:
+// 1. L3/L4
+// 2. L4-only (wildcard L3)
+// 3. L3-only (wildcard L4)
+// 4. Allow-all
+//
+// This means that the L4-only allow visibility key can only be added if there is an
+// allow-all key, and all L3-only deny keys are expanded to L3/L4 keys. If no
+// L4-only key is added then also the L3-only allow keys need to be expanded to
+// L3/L4 keys for visibility redirection. In addition the existing L3/L4 and L4-only
+// allow keys need to be redirected to the proxy port, if not already redirected.
+//
+// The above can be accomplished by:
+//
+// 1. Change existing L4-only ALLOW key on matching port that does not already
+//    redirect to redirect.
+//    - e.g., 0:80=allow,0 -> 0:80=allow,<proxyport>
+// 2. If allow-all policy exists, add L4-only visibility redirect key if the L4-only
+//    key does not already exist.
+//    - e.g., 0:0=allow,0 -> add 0:80=allow,<proxyport> if 0:80 does not exist
+//      - this allows all traffic on port 80, but see step 5 below.
+// 3. Change all L3/L4 ALLOW keys on matching port that do not already redirect to
+//    redirect.
+//    - e.g, <ID1>:80=allow,0 -> <ID1>:80=allow,<proxyport>
+// 4. For each L3-only ALLOW key add the corresponding L3/L4 ALLOW redirect if no
+//    L3/L4 key already exists and no L4-only key already exists and one is not added.
+//    - e.g., <ID2>:0=allow,0 -> add <ID2>:80=allow,<proxyport> if <ID2>:80
+//      and 0:80 do not exist
+// 5. If a new L4-only key was added: For each L3-only DENY key add the
+//    corresponding L3/L4 DENY key if no L3/L4 key already exists.
+//    - e.g., <ID3>:0=deny,0 -> add <ID3>:80=deny,0 if <ID3>:80 does not exist
+//
+// With the above we only change/expand existing allow keys to redirect, and
+// expand existing drop keys to also drop on the port of interest, if a new
+// L4-only key allowing the port is added.
+//
+// 'adds' and 'deletes' are updates with the changes made. 'adds' contains the new values
```

Comment:
```suggestion
// 'adds' and 'deletes' are updated with the changes made. 'adds' contains the new values
```

Any particular reason to pass in these output parameters rather than just returning them? Is it just to avoid the merging logic in the caller?

Reply:
Yes. All these changes are collected into `MapStateChanges` so it is logical to pass them around as these maps exist independent of any one call being made.

Reply:
Fixed typo.